File: src/main/java/com/cse3063f25grp1/retrieval/KeywordRetriever.java
package com.cse3063f25grp1.retrieval;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.cse3063f25grp1.data.ChunkStore;
import com.cse3063f25grp1.model.Chunk;
import com.cse3063f25grp1.model.Hit;

public class KeywordRetriever implements Retriever {

    private final int topK;

    public KeywordRetriever(int topK) {
        this.topK = topK;
    }

    @Override
    public List<Hit> retrieve(List<String> queryTerms, ChunkStore store) {
        if (queryTerms == null || queryTerms.isEmpty())
            return Collections.emptyList();

        Map<String, Hit> hitMap = new HashMap<>();

        for (Chunk chunk : store.getAllChunks()) {
            String chunkTextLower = safeLower(chunk.getText());
            int totalTf = 0;

            for (String term : queryTerms) {
                if (term == null || term.isEmpty())
                    continue;
                totalTf += countOccurrences(chunkTextLower, term.toLowerCase());
            }

            if (totalTf > 0) {
                String key = chunk.getDocId() + "||" + chunk.getChunkId();
                Hit h = hitMap.get(key);
                if (h == null) {
                    h = new Hit(chunk.getDocId(), chunk.getChunkId(), totalTf);
                    hitMap.put(key, h);
                } else {
                    h.setScore(h.getScore() + totalTf);
                }
            }
        }

        List<Hit> hits = new ArrayList<>(hitMap.values());
        hits.sort(Comparator
                .comparingInt(Hit::getScore).reversed()
                .thenComparing(Hit::getDocId)
                .thenComparing(Hit::getChunkId));

        if (hits.size() > topK) {
            return hits.subList(0, topK);
        } else {
            return hits;
        }
    }

    private static int countOccurrences(String haystackLower, String needleLower) {
        if (haystackLower == null || needleLower == null)
            return 0;
        int count = 0;
        int idx = haystackLower.indexOf(needleLower);
        while (idx != -1) {
            count++;
            idx = haystackLower.indexOf(needleLower, idx + needleLower.length());
        }
        return count;
    }

    private static String safeLower(String s) {
        return s == null ? "" : s.toLowerCase();
    }

}

File: iteration1_py/src/retrieval/keyword_retriever.py
from typing import List, Dict

from data.chunk_store import ChunkStore
from model.chunk import Chunk
from model.hit import Hit
from retriever_abstract import Retriever 

class KeywordRetriever(Retriever):

    def __init__(self, top_k: int):
        self.top_k = top_k

    def _safe_lower(self, s: str) -> str:
        return s.lower() if s is not None else ""

    def _count_occurrences(self, haystack_lower: str, needle_lower: str) -> int:
        if not haystack_lower or not needle_lower:
            return 0
        return haystack_lower.count(needle_lower)

    def retrieve(self, query_terms: List[str], store) -> List: 
        if not query_terms:
            return []

        hit_map: Dict[str, object] = {}

        for chunk in store.get_all_chunks():
            chunk_text_lower = self._safe_lower(chunk.get_text())
            total_tf = 0

            for term in query_terms:
                if not term:
                    continue
                total_tf += self._count_occurrences(chunk_text_lower, term.lower())

            if total_tf > 0:
                key = f"{chunk.get_doc_id()}||{chunk.get_chunk_id()}"
                
                h = hit_map.get(key)
                
                if h is None:
                    h = object(chunk.get_doc_id(), chunk.get_chunk_id(), total_tf) 
                else:
                    h.set_score(h.get_score() + total_tf)

        hits: List[object] = list(hit_map.values())

        hits.sort(key=lambda h: (
            -h.get_score(),           
            h.get_doc_id(),           
            h.get_chunk_id()
        ))
        
        return hits[:self.top_k]